index.js

```diff
@@ -7,7 +7,7 @@ const flatten = source => (start, sink) => {
     if (t === 1) (innerTalkback || outerTalkback)(1, d);
     if (t === 2) {
       innerTalkback && innerTalkback(2);
-      outerTalkback(2);
+      if (!outerEnded) outerTalkback(2);
```

Comment:
Alternatively, we could remove `outerEnded` completely, just clear `outerTalkback` and use its presence instead:
```diff
diff --git a/index.js b/index.js
index 1e2bbb6..1827cc5 100644
--- a/index.js
+++ b/index.js
@@ -1,13 +1,12 @@
 const flatten = source => (start, sink) => {
   if (start !== 0) return;
-  let outerEnded = false;
   let outerTalkback;
   let innerTalkback;
   function talkback(t, d) {
     if (t === 1) (innerTalkback || outerTalkback)(1, d);
     if (t === 2) {
       innerTalkback && innerTalkback(2);
-      if (!outerEnded) outerTalkback(2);
+      outerTalkback && outerTalkback(2);
     }
   }
   source(0, (T, D) => {
@@ -26,7 +25,7 @@ const flatten = source => (start, sink) => {
           outerTalkback(2);
           sink(2, d);
         } else if (t === 2) {
-          if (outerEnded) sink(2);
+          if (!outerTalkback) sink(2);
           else {
             innerTalkback = void 0;
             outerTalkback(1);
@@ -38,7 +37,7 @@ const flatten = source => (start, sink) => {
       sink(2, D);
     } else if (T === 2) {
       if (!innerTalkback) sink(2);
-      else outerEnded = true;
+      else outerTalkback = void 0;
     }
   });
 };
```